File: app/src/main/java/com/tunjid/androidbootstrap/model/Tile.java
package com.tunjid.androidbootstrap.model;

import android.graphics.Color;

import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;

public class Tile implements Differentiable {

    private static final int[] colors = {
            Color.RED,
            Color.BLACK,
            Color.BLUE,
            Color.CYAN,
            Color.MAGENTA,
            Color.GREEN,
            Color.LTGRAY,
            Color.DKGRAY
    };

    private final int number;
    @ColorInt
    private final int color;

    private Tile(int number, int color) {
        this.number = number;
        this.color = color;
    }

    public static Tile generate(int id) {
        return new Tile(id, colors[randomIndex()]);
    }

    public int getNumber() {
        return number;
    }

    @ColorInt
    public int getColor() {
        return color;
    }

    @Override
    public String getId() {
        return String.valueOf(number);
    }

    @Override
    public boolean areContentsTheSame(Differentiable other) {
        if (!(other instanceof Tile)) return false;

        Tile otherTile = (Tile) other;
        return getId().equals(otherTile.getId()) && getColor() == otherTile.getColor();
    }

    @Override
    public Object getChangePayload(Differentiable other) {
        return other;
    }

    @NonNull
    @Override
    public String toString() {
        return getId();
    }

    private static int randomIndex() {
        return (int) (Math.random() * colors.length);
    }
}

File: app/src/main/java/com/tunjid/androidbootstrap/model/Tile.kt
package com.tunjid.androidbootstrap.model

import android.graphics.Color

import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable

import androidx.annotation.ColorInt

class Tile private constructor(
        val number: Int,
        @field:ColorInt @get:ColorInt val color: Int) : Differentiable {

    override fun getId(): String = number.toString()

    override fun areContentsTheSame(other: Differentiable): Boolean =
            if (other !is Tile) false else id == other.id && color == other.color

    override fun getChangePayload(other: Differentiable): Any = other

    override fun toString(): String = id

    companion object {

        private val colors = intArrayOf(Color.RED, Color.BLACK, Color.BLUE, Color.CYAN, Color.MAGENTA, Color.GREEN, Color.LTGRAY, Color.DKGRAY)

        fun generate(id: Int): Tile = Tile(id, colors[randomIndex()])

        private fun randomIndex(): Int = (Math.random() * colors.size).toInt()
    }
}
